File: android/app/src/main/java/com/example/mycarnavi/MapViewModel.kt
package com.example.mycarnavi

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class MapViewModel : ViewModel() {

    private val routesApiClient = RoutesApiClient(BuildConfig.MAPS_API_KEY)

    private val placesApiClient = PlacesApiClient(BuildConfig.MAPS_API_KEY)

    private val _camera = MutableStateFlow(
        CameraState(
            latitude = 35.681236,
            longitude = 139.767125,
            zoom = 15f
        )
    )

    val camera: StateFlow<CameraState> = _camera

    private val _destination = MutableStateFlow<Destination?>(null)

    val destination: StateFlow<Destination?> = _destination

    fun setDestinationToCenter() {
        val camera = _camera.value
        setDestination(camera.latitude, camera.longitude)
    }

    fun setDestination(latitude: Double, longitude: Double) {
        _destination.value = Destination(
            latitude = latitude,
            longitude = longitude
        )
        // 目的地を画面中心に表示する
        _camera.value = _camera.value.copy(
            latitude = latitude,
            longitude = longitude
        )
        requestRoute()
    }

    private val _route = MutableStateFlow<RouteInfo?>(null)

    val route: StateFlow<RouteInfo?> = _route

    private val _parkingPlaces = MutableStateFlow<List<ParkingPlace>>(emptyList())

    val parkingPlaces: StateFlow<List<ParkingPlace>> = _parkingPlaces

    /** 駐車場マーカーの表示切り替え。非表示なら検索して表示、表示中なら消す */
    fun toggleParkingSearch() {
        if (_parkingPlaces.value.isNotEmpty()) {
            _parkingPlaces.value = emptyList()
            return
        }

        val camera = _camera.value

        viewModelScope.launch(Dispatchers.IO) {
            _parkingPlaces.value = try {
                placesApiClient.searchParking(
                    centerLatitude = camera.latitude,
                    centerLongitude = camera.longitude
                )
            } catch (e: Exception) {
                Log.e("MapViewModel", "Failed to search parking", e)
                emptyList()
            }
        }
    }

    private fun requestRoute() {
        val origin = _currentLocation.value
        val destination = _destination.value

        if (origin == null || destination == null) {
            _route.value = null
            return
        }

        viewModelScope.launch(Dispatchers.IO) {
            _route.value = try {
                routesApiClient.fetchRoute(
                    originLatitude = origin.latitude,
                    originLongitude = origin.longitude,
                    destinationLatitude = destination.latitude,
                    destinationLongitude = destination.longitude
                )
            } catch (e: Exception) {
                Log.e("MapViewModel", "Failed to fetch route", e)
                null
            }
        }
    }

    private val _currentLocation = MutableStateFlow<CurrentLocation?>(null)

    val currentLocation: StateFlow<CurrentLocation?> = _currentLocation

    private val _locationPermissionGranted = MutableStateFlow(false)

    val locationPermissionGranted: StateFlow<Boolean> = _locationPermissionGranted

    fun setLocationPermissionGranted(granted: Boolean) {
        _locationPermissionGranted.value = granted
    }

    fun updateCurrentLocation(latitude: Double, longitude: Double) {
        _currentLocation.value = CurrentLocation(latitude, longitude)
    }

    fun centerOnCurrentLocation() {
        val location = _currentLocation.value ?: return
        _camera.value = _camera.value.copy(
            latitude = location.latitude,
            longitude = location.longitude
        )
    }

    fun moveWest() {
        move(-200.0, 0.0)
    }

    fun moveEast() {
        move(200.0, 0.0)
    }

    fun moveNorth() {
        move(0.0, 200.0)
    }

    fun moveSouth() {
        move(0.0, -200.0)
    }

    fun zoomIn() {
        _camera.value = _camera.value.copy(
            zoom = (_camera.value.zoom + 1f).coerceAtMost(21f)
        )
    }

    fun zoomOut() {
        _camera.value = _camera.value.copy(
            zoom = (_camera.value.zoom - 1f).coerceAtLeast(3f)
        )
    }

    private fun move(eastMeters: Double, northMeters: Double) {

        val camera = _camera.value

        // ズームが1段階下がると表示範囲が約2倍になるため、
        // 移動量も2倍にして画面上の移動量を一定に保つ（zoom=15で基準の200m）
        val scale = Math.pow(2.0, (15.0 - camera.zoom))

        val dLat = northMeters * scale / 111320.0

        val dLon = eastMeters * scale /
                (111320.0 * kotlin.math.cos(Math.toRadians(camera.latitude)))

        _camera.value = camera.copy(
            latitude = camera.latitude + dLat,
            longitude = camera.longitude + dLon
        )
    }
}

File: ios/Navi/ViewModels/MapViewModel.swift
import CoreLocation
import Foundation
import Observation

@MainActor
@Observable
final class MapViewModel {
    var camera = CameraState(latitude: 35.681236, longitude: 139.767125, zoom: 15)
    var destination: Destination?
    var currentLocation: CurrentLocation?
    var locationPermissionGranted = false
    var route: RouteInfo?
    var parkingPlaces: [ParkingPlace] = []
    var statusMessage: String?

    var mapProxy: MapViewProxy?

    private let routesApiClient = RoutesApiClient()
    private let placesApiClient = PlacesApiClient()
    private var panTask: Task<Void, Never>?

    func setLocationPermissionGranted(_ granted: Bool) {
        locationPermissionGranted = granted
    }

    func updateCurrentLocation(latitude: Double, longitude: Double) {
        let wasMissing = currentLocation == nil
        currentLocation = CurrentLocation(latitude: latitude, longitude: longitude)
        if destination != nil && wasMissing {
            requestRoute()
        }
    }

    func setDestinationToCenter() {
        if let center = mapProxy?.visibleCenter {
            setDestination(latitude: center.latitude, longitude: center.longitude)
        } else {
            setDestination(latitude: camera.latitude, longitude: camera.longitude)
        }
    }

    func setDestination(latitude: Double, longitude: Double) {
        destination = Destination(latitude: latitude, longitude: longitude)
        camera = CameraState(latitude: latitude, longitude: longitude, zoom: camera.zoom)
        statusMessage = "目的地を設定しました"
        requestRoute()
    }

    func centerOnCurrentLocation() {
        guard let location = currentLocation else {
            statusMessage = "現在地を取得できません"
            return
        }
        camera = CameraState(
            latitude: location.latitude,
            longitude: location.longitude,
            zoom: camera.zoom
        )
    }

    func toggleParkingSearch() {
        if !parkingPlaces.isEmpty {
            parkingPlaces = []
            return
        }

        let snapshot = camera
        Task {
            parkingPlaces = await placesApiClient.searchParking(
                centerLatitude: snapshot.latitude,
                centerLongitude: snapshot.longitude
            )
        }
    }

    func startPan(_ direction: PanDirection) {
        panTask?.cancel()
        panTask = Task {
            while !Task.isCancelled {
                switch direction {
                case .west: moveWest()
                case .east: moveEast()
                case .north: moveNorth()
                case .south: moveSouth()
                }
                try? await Task.sleep(for: .milliseconds(120))
            }
        }
    }

    func stopPan() {
        panTask?.cancel()
        panTask = nil
    }

    func zoomIn() {
        camera = CameraState(
            latitude: camera.latitude,
            longitude: camera.longitude,
            zoom: min(camera.zoom + 1, 21)
        )
    }

    func zoomOut() {
        camera = CameraState(
            latitude: camera.latitude,
            longitude: camera.longitude,
            zoom: max(camera.zoom - 1, 3)
        )
    }

    func moveWest() { move(eastMeters: -200, northMeters: 0) }
    func moveEast() { move(eastMeters: 200, northMeters: 0) }
    func moveNorth() { move(eastMeters: 0, northMeters: 200) }
    func moveSouth() { move(eastMeters: 0, northMeters: -200) }

    private func move(eastMeters: Double, northMeters: Double) {
        let scale = pow(2.0, 15.0 - Double(camera.zoom))
        let dLat = northMeters * scale / 111_320.0
        let dLon = eastMeters * scale / (111_320.0 * cos(camera.latitude * .pi / 180))
        camera = CameraState(
            latitude: camera.latitude + dLat,
            longitude: camera.longitude + dLon,
            zoom: camera.zoom
        )
    }

    private func requestRoute() {
        guard let destination else {
            route = nil
            return
        }

        guard let origin = currentLocation else {
            route = nil
            statusMessage = "ルート表示には現在地（位置情報）が必要です"
            return
        }

        Task {
            let fetched = await routesApiClient.fetchRoute(
                originLatitude: origin.latitude,
                originLongitude: origin.longitude,
                destinationLatitude: destination.latitude,
                destinationLongitude: destination.longitude
            )
            route = fetched
            if fetched == nil {
                statusMessage = "ルートを取得できませんでした"
            }
        }
    }
}
